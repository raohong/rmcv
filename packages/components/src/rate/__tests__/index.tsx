import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';
import { Rate, rateClassNames } from '..';

const testId = 'rate';

it('render correctly', () => {
  const tree = render(<Rate />);

  expect(tree.asFragment()).toMatchSnapshot();
});

it('render with defaultValue', () => {
  render(<Rate defaultValue={2} data-testid={testId} />);

  expect(
    screen.getByTestId(testId).querySelectorAll(`.${rateClassNames.fullIcon}`)
      .length,
  ).toBe(2);
});

it('render with onChange', async () => {
  const onChange = jest.fn();

  render(<Rate onChange={onChange} data-testid={testId} />);

  const second = screen
    .getByTestId(testId)
    .querySelector(`.${rateClassNames.item}:nth-child(2)`)!;

  await userEvent.click(second);

  expect(onChange).toHaveBeenCalledWith(2);
});

it('render with value', async () => {
  const App = () => {
    const [value, setValue] = useState(0);

    return <Rate value={value} onChange={setValue} data-testid={testId} />;
  };
  render(<App />);

  const second = screen
    .getByTestId(testId)
    .querySelector(`.${rateClassNames.item}:nth-child(2)`)!;

  await userEvent.click(second);

  expect(second.querySelector(`.${rateClassNames.mask}`)).not.toBeNull();
});

it('render with allowHalf', () => {
  render(<Rate defaultValue={1.5} data-testid={testId} allowHalf />);

  const target = screen
    .getByTestId(testId)
    .querySelector(`.${rateClassNames.item}:nth-child(2) .${rateClassNames.mask}`);

  expect(target).toHaveStyle({ 'clip-path': 'inset(0 50% 0 0)' });
});

it('render with readonly', async () => {
  const onChange = jest.fn();
  render(<Rate onChange={onChange} data-testid={testId} readonly />);

  const second = screen
    .getByTestId(testId)
    .querySelector(`.${rateClassNames.item}:nth-child(2)`)!;

  await userEvent.click(second);
  expect(onChange).not.toHaveBeenCalled();
  expect(screen.getByTestId(testId)).toHaveClass(rateClassNames.readonly);
});

it('render with disabled', async () => {
  const onChange = jest.fn();
  render(<Rate onChange={onChange} data-testid={testId} disabled />);

  const second = screen
    .getByTestId(testId)
    .querySelector(`.${rateClassNames.item}:nth-child(2)`)!;

  await userEvent.click(second);
  expect(onChange).not.toHaveBeenCalled();
  expect(screen.getByTestId(testId)).toHaveClass(rateClassNames.disabled);
});

it('render decimal value', () => {
  render(<Rate data-testid={testId} defaultValue={1.2} readonly allowHalf />);

  const target = screen
    .getByTestId(testId)
    .querySelector(`.${rateClassNames.item}:nth-child(2) .${rateClassNames.mask}`);

  expect(target).toHaveStyle({ 'clip-path': 'inset(0 80% 0 0)' });
});
