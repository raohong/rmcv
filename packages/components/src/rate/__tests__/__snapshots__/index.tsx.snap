// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  color: #c8c9cc;
  font-size: 20px;
}

.emotion-1 {
  box-sizing: border-box;
  position: relative;
  width: 1em;
  height: 1em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  box-sizing: border-box;
  width: 1em;
  height: 1em;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-3 {
  display: inline-block;
  color: inherit;
  font-style: normal;
  line-height: 0;
  box-sizing: border-box;
  text-rendering: optimizeLegibility;
  -webkit-font-smooth: antialiased;
}

.emotion-3 * {
  line-height: 1;
}

<div
    aria-disabled="false"
    aria-readonly="false"
    class="RVuiRate-root emotion-0"
    data-value="0"
  >
    <div
      class="RVuiRate-item emotion-1"
      role="button"
      tabindex="0"
    >
      <span
        class="RVuiRate-icon emotion-2"
      >
        <span
          aria-label="Star-Outlined"
          class="RVuiIcon-root emotion-3"
          role="img"
        >
          <svg
            fill="currentColor"
            height="1em"
            version="1.1"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M720.692224 636.6976l170.748928-191.096832-250.462208-54.23616L512 169.920512 383.021056 391.364608l-250.462208 54.23616 170.748928 191.096832-25.81504 254.963712L512 788.32128l234.507264 103.340032-25.81504-254.963712zM512 850.48832L255.744 963.412992c-14.37696 6.335488-31.16544-0.183296-37.499904-14.558208a28.444672 28.444672 0 0 1-2.270208-14.336l28.208128-278.60992L57.6 447.08864c-10.468352-11.71456-9.45664-29.696 2.25792-40.163328a28.444672 28.444672 0 0 1 12.932096-6.588416l273.690624-59.267072 140.941312-241.98144c7.906304-13.574144 25.320448-18.168832 38.895616-10.262528a28.444672 28.444672 0 0 1 10.262528 10.262528l140.941312 241.98144 273.690624 59.267072c15.353856 3.323904 25.105408 18.465792 21.78048 33.819648a28.444672 28.444672 0 0 1-6.58944 12.93312L779.81696 655.906816l28.208128 278.60992c1.583104 15.629312-9.8048 29.58336-25.434112 31.16544-4.89472 0.495616-9.833472-0.28672-14.336-2.271232L512 850.489344z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="RVuiRate-item emotion-1"
      role="button"
      tabindex="0"
    >
      <span
        class="RVuiRate-icon emotion-2"
      >
        <span
          aria-label="Star-Outlined"
          class="RVuiIcon-root emotion-3"
          role="img"
        >
          <svg
            fill="currentColor"
            height="1em"
            version="1.1"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M720.692224 636.6976l170.748928-191.096832-250.462208-54.23616L512 169.920512 383.021056 391.364608l-250.462208 54.23616 170.748928 191.096832-25.81504 254.963712L512 788.32128l234.507264 103.340032-25.81504-254.963712zM512 850.48832L255.744 963.412992c-14.37696 6.335488-31.16544-0.183296-37.499904-14.558208a28.444672 28.444672 0 0 1-2.270208-14.336l28.208128-278.60992L57.6 447.08864c-10.468352-11.71456-9.45664-29.696 2.25792-40.163328a28.444672 28.444672 0 0 1 12.932096-6.588416l273.690624-59.267072 140.941312-241.98144c7.906304-13.574144 25.320448-18.168832 38.895616-10.262528a28.444672 28.444672 0 0 1 10.262528 10.262528l140.941312 241.98144 273.690624 59.267072c15.353856 3.323904 25.105408 18.465792 21.78048 33.819648a28.444672 28.444672 0 0 1-6.58944 12.93312L779.81696 655.906816l28.208128 278.60992c1.583104 15.629312-9.8048 29.58336-25.434112 31.16544-4.89472 0.495616-9.833472-0.28672-14.336-2.271232L512 850.489344z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="RVuiRate-item emotion-1"
      role="button"
      tabindex="0"
    >
      <span
        class="RVuiRate-icon emotion-2"
      >
        <span
          aria-label="Star-Outlined"
          class="RVuiIcon-root emotion-3"
          role="img"
        >
          <svg
            fill="currentColor"
            height="1em"
            version="1.1"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M720.692224 636.6976l170.748928-191.096832-250.462208-54.23616L512 169.920512 383.021056 391.364608l-250.462208 54.23616 170.748928 191.096832-25.81504 254.963712L512 788.32128l234.507264 103.340032-25.81504-254.963712zM512 850.48832L255.744 963.412992c-14.37696 6.335488-31.16544-0.183296-37.499904-14.558208a28.444672 28.444672 0 0 1-2.270208-14.336l28.208128-278.60992L57.6 447.08864c-10.468352-11.71456-9.45664-29.696 2.25792-40.163328a28.444672 28.444672 0 0 1 12.932096-6.588416l273.690624-59.267072 140.941312-241.98144c7.906304-13.574144 25.320448-18.168832 38.895616-10.262528a28.444672 28.444672 0 0 1 10.262528 10.262528l140.941312 241.98144 273.690624 59.267072c15.353856 3.323904 25.105408 18.465792 21.78048 33.819648a28.444672 28.444672 0 0 1-6.58944 12.93312L779.81696 655.906816l28.208128 278.60992c1.583104 15.629312-9.8048 29.58336-25.434112 31.16544-4.89472 0.495616-9.833472-0.28672-14.336-2.271232L512 850.489344z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="RVuiRate-item emotion-1"
      role="button"
      tabindex="0"
    >
      <span
        class="RVuiRate-icon emotion-2"
      >
        <span
          aria-label="Star-Outlined"
          class="RVuiIcon-root emotion-3"
          role="img"
        >
          <svg
            fill="currentColor"
            height="1em"
            version="1.1"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M720.692224 636.6976l170.748928-191.096832-250.462208-54.23616L512 169.920512 383.021056 391.364608l-250.462208 54.23616 170.748928 191.096832-25.81504 254.963712L512 788.32128l234.507264 103.340032-25.81504-254.963712zM512 850.48832L255.744 963.412992c-14.37696 6.335488-31.16544-0.183296-37.499904-14.558208a28.444672 28.444672 0 0 1-2.270208-14.336l28.208128-278.60992L57.6 447.08864c-10.468352-11.71456-9.45664-29.696 2.25792-40.163328a28.444672 28.444672 0 0 1 12.932096-6.588416l273.690624-59.267072 140.941312-241.98144c7.906304-13.574144 25.320448-18.168832 38.895616-10.262528a28.444672 28.444672 0 0 1 10.262528 10.262528l140.941312 241.98144 273.690624 59.267072c15.353856 3.323904 25.105408 18.465792 21.78048 33.819648a28.444672 28.444672 0 0 1-6.58944 12.93312L779.81696 655.906816l28.208128 278.60992c1.583104 15.629312-9.8048 29.58336-25.434112 31.16544-4.89472 0.495616-9.833472-0.28672-14.336-2.271232L512 850.489344z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="RVuiRate-item emotion-1"
      role="button"
      tabindex="0"
    >
      <span
        class="RVuiRate-icon emotion-2"
      >
        <span
          aria-label="Star-Outlined"
          class="RVuiIcon-root emotion-3"
          role="img"
        >
          <svg
            fill="currentColor"
            height="1em"
            version="1.1"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M720.692224 636.6976l170.748928-191.096832-250.462208-54.23616L512 169.920512 383.021056 391.364608l-250.462208 54.23616 170.748928 191.096832-25.81504 254.963712L512 788.32128l234.507264 103.340032-25.81504-254.963712zM512 850.48832L255.744 963.412992c-14.37696 6.335488-31.16544-0.183296-37.499904-14.558208a28.444672 28.444672 0 0 1-2.270208-14.336l28.208128-278.60992L57.6 447.08864c-10.468352-11.71456-9.45664-29.696 2.25792-40.163328a28.444672 28.444672 0 0 1 12.932096-6.588416l273.690624-59.267072 140.941312-241.98144c7.906304-13.574144 25.320448-18.168832 38.895616-10.262528a28.444672 28.444672 0 0 1 10.262528 10.262528l140.941312 241.98144 273.690624 59.267072c15.353856 3.323904 25.105408 18.465792 21.78048 33.819648a28.444672 28.444672 0 0 1-6.58944 12.93312L779.81696 655.906816l28.208128 278.60992c1.583104 15.629312-9.8048 29.58336-25.434112 31.16544-4.89472 0.495616-9.833472-0.28672-14.336-2.271232L512 850.489344z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
